src/rtc.c

```diff
@@ -0,0 +1,518 @@
+ * write to actually complete on real hardware. It is recommended to wait
+ * using #JOYBUS_RTC_WRITE_DELAY before setting the RTC control block back
+ * to #JOYBUS_RTC_CONTROL_MODE_RUN after setting the time.
+ *
+ * Generally, you should not need to call this function directly. Prefer
+ * calling the high-level #joybus_rtc_set function which handles the RTC
+ * control block writes, delays, and detects ignored writes.
+ *
+ * @param[in]  rtc_time
+ *             Source pointer for the RTC time data structure
+ *
+ * @return the status of the real-time clock
+ */
+void joybus_rtc_write_time( const rtc_time_t * rtc_time )
+{
+    uint8_t data[sizeof(uint64_t)];
+
+    int year = rtc_time->year - 1900;
+    data[0] = byte_to_bcd(rtc_time->sec);
+    data[1] = byte_to_bcd(rtc_time->min);
+    data[2] = byte_to_bcd(rtc_time->hour) + 0x80;
+    data[3] = byte_to_bcd(rtc_time->day);
+    data[4] = byte_to_bcd(rtc_time->week_day);
+    data[5] = byte_to_bcd(rtc_time->month + 1);
+    data[6] = byte_to_bcd(year);
+    data[7] = byte_to_bcd(year / 100);
+
+    joybus_rtc_write( 2, (uint64_t *)data );
+}
+
+/**
+ * @brief High-level convenience helper to prepare the RTC after startup.
+ *
+ * This function is a good idea to call immediately after verifying
+ * that the RTC is present and before the first call to
+ * #joybus_rtc_read_time so that you can be sure the RTC is ready.
+ *
+ * This function will take approximately 25 milliseconds to complete.
+ *
+ * It is not necessary to call this after calling #joybus_rtc_set.
+ */
+void joybus_rtc_run( void )
+{
+    uint32_t calibration;
+    /* Read the calibration data from the control block */
+    joybus_rtc_read_control( NULL, &calibration );
+    /* Put the RTC into normal operating mode */
+    joybus_rtc_write_control( JOYBUS_RTC_CONTROL_MODE_RUN, calibration );
+    wait_ms( JOYBUS_RTC_WRITE_DELAY );
+}
+
+/**
+ * @brief High-level convenience helper to set the RTC date/time.
+ *
+ * Prepares the RTC for writing, sets the new time, and resumes the clock.
+ *
+ * This process will take approximately 70 milliseconds to complete.
+ * Unfortunately, the best way to ensure that writes to the RTC have
+ * actually finished is by waiting for a fixed duration. Emulators may not
+ * accurately reflect this, but this delay is necessary on real hardware.
+ *
+ * @param[in]   rtc_time
+ *              Source pointer for the RTC time data structure
+ */
+void joybus_rtc_set( const rtc_time_t * write_time )
+{
+    uint32_t calibration;
+    /* Read the calibration data from the control block */
+    joybus_rtc_read_control( NULL, &calibration );
+    /* Prepare the RTC to write the time, preserving the calibration data */
+    joybus_rtc_write_control( JOYBUS_RTC_CONTROL_MODE_SET, calibration );
+    wait_ms( JOYBUS_RTC_WRITE_DELAY );
+    /* Check the RTC status to make sure RTC block 2 writes are supported */
+    if( joybus_rtc_is_busy() )
```

Comment:
What happens if this if is not taken? It looks to me the function fails, but it's not reported. Also the RTC is left in write mode?

Reply:
Some emulators (Mupen64Plus and cen64 [without my patch]) don't implement the RTC busy status at all and also don't support writes. This just bails on the write if the busy status is unimplemented.

The reason it fails silently is because I have to treat all RTC write errors as silent failures. The intent is that the developer calls `joybus_rtc_is_writable` to determine if writes are implemented. Somewhere in the git history I had this function return a bool of whether the write was successful, but it ended up basically being an implementation of `joybus_rtc_is_writable` inside `joybus_rtc_set`.

I'm open to suggestions about how to handle these silent writes better

Reply:
Addressed in https://github.com/DragonMinded/libdragon/pull/152/commits/e46fbb07a6cbb01a98f05a2d3b170d4ff1860b77